File: android/src/main/java/com/reactnativeviewer/rnvnetwork/RNVNetworkDevModule.kt
package com.reactnativeviewer.rnvnetwork

import android.os.Build
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.ReadableType
import com.facebook.react.modules.core.DeviceEventManagerModule
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID

class RNVNetworkDevModule(
    reactContext: ReactApplicationContext
) : ReactContextBaseJavaModule(reactContext) {
    companion object {
        const val MODULE_NAME = "RNVNetworkDevModule"
        private const val STATUS_EVENT_NAME = "RNVNetworkDevStatus"
        private const val SDK_VERSION = "0.1.0"
    }

    private val client = RNVNetworkDevWebSocketClient()
    private var enabled = true
    private val sessionIdentifier = UUID.randomUUID().toString()
    private var listenerCount = 0

    init {
        client.statusHandler = { state, detail ->
            emitStatus(state, detail)
        }
    }

    override fun getName(): String = MODULE_NAME

    override fun getConstants(): MutableMap<String, Any> {
        return mutableMapOf(
            "statusEventName" to STATUS_EVENT_NAME,
            "sdkVersion" to SDK_VERSION
        )
    }

    @ReactMethod
    fun configure(options: ReadableMap?) {
        enabled = options?.getOptionalBoolean("enabled") ?: true

        if (!enabled) {
            client.disconnect()
            emitStatus("disabled", "SDK disabled by configuration.")
            return
        }

        val viewerURL = options?.getOptionalString("viewerURL")
        val maxQueueSize = options?.getOptionalInt("maxQueueSize") ?: 200
        val connectionHeaders = options?.getOptionalMap("connectionHeaders")?.toStringMap() ?: emptyMap()

        client.configure(
            urlString = viewerURL,
            maxQueueSize = maxQueueSize,
            connectionHeaders = connectionHeaders
        )
    }

    @ReactMethod
    fun setEnabled(enabled: Boolean) {
        this.enabled = enabled
        if (!enabled) {
            client.disconnect()
            emitStatus("disabled", "SDK disabled at runtime.")
        }
    }

    @ReactMethod
    fun captureEvent(event: ReadableMap?) {
        if (event == null) {
            return
        }

        val events = JSONArray().put(event.toJSONObject())
        enqueueEvents(events)
    }

    @ReactMethod
    fun captureEvents(events: ReadableArray?) {
        if (events == null || events.size() == 0) {
            return
        }

        val array = JSONArray()
        for (index in 0 until events.size()) {
            val event = events.getMap(index) ?: continue
            array.put(event.toJSONObject())
        }

        enqueueEvents(array)
    }

    @ReactMethod
    fun addListener(eventName: String) {
        listenerCount += 1
    }

    @ReactMethod
    fun removeListeners(count: Int) {
        listenerCount = maxOf(0, listenerCount - count)
    }

    override fun invalidate() {
        client.disconnect()
        super.invalidate()
    }

    private fun enqueueEvents(events: JSONArray) {
        if (!enabled || events.length() == 0) {
            return
        }

        client.enqueueEnvelope(buildEnvelope(events))
    }

    private fun buildEnvelope(events: JSONArray): JSONObject {
        return JSONObject()
            .put(
                "sdk",
                JSONObject()
                    .put("name", "rnv_network_sdk_android")
                    .put("version", SDK_VERSION)
                    .put("schemaVersion", 1)
            )
            .put("session", sessionMetadata())
            .put("events", events)
    }

    private fun sessionMetadata(): JSONObject {
        val packageName = reactApplicationContext.packageName
        val packageManager = reactApplicationContext.packageManager
        val applicationInfo = reactApplicationContext.applicationInfo
        val appName = packageManager.getApplicationLabel(applicationInfo)?.toString().orEmpty()

        return JSONObject()
            .put("id", sessionIdentifier)
            .put("platform", "android")
            .put("bundleIdentifier", packageName)
            .put("appName", appName)
            .put("deviceName", deviceName())
            .put("systemName", "Android")
            .put("systemVersion", Build.VERSION.RELEASE ?: "")
            .put("isSimulator", isEmulator())
    }

    private fun emitStatus(state: String, detail: String?) {
        if (listenerCount <= 0 || !reactApplicationContext.hasActiveCatalystInstance()) {
            return
        }

        val payload = Arguments.createMap().apply {
            putString("state", state)
            putString("sessionId", sessionIdentifier)
            if (!detail.isNullOrBlank()) {
                putString("detail", detail)
            }
        }

        reactApplicationContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit(STATUS_EVENT_NAME, payload)
    }

    private fun deviceName(): String {
        val manufacturer = Build.MANUFACTURER?.trim().orEmpty()
        val model = Build.MODEL?.trim().orEmpty()

        return when {
            manufacturer.isEmpty() -> model
            model.startsWith(manufacturer, ignoreCase = true) -> model
            model.isEmpty() -> manufacturer
            else -> "$manufacturer $model"
        }
    }

    private fun isEmulator(): Boolean {
        return (
            Build.FINGERPRINT.startsWith("generic") ||
                Build.FINGERPRINT.startsWith("unknown") ||
                Build.MODEL.contains("google_sdk") ||
                Build.MODEL.contains("Emulator") ||
                Build.MODEL.contains("Android SDK built for x86") ||
                Build.BRAND.startsWith("generic") && Build.DEVICE.startsWith("generic") ||
                "google_sdk" == Build.PRODUCT
            )
    }
}

private fun ReadableMap.getOptionalBoolean(key: String): Boolean? {
    if (!hasKey(key) || isNull(key)) {
        return null
    }
    return getBoolean(key)
}

private fun ReadableMap.getOptionalString(key: String): String? {
    if (!hasKey(key) || isNull(key)) {
        return null
    }
    return getString(key)
}

private fun ReadableMap.getOptionalInt(key: String): Int? {
    if (!hasKey(key) || isNull(key)) {
        return null
    }

    return when (getType(key)) {
        ReadableType.Number -> getDouble(key).toInt()
        else -> null
    }
}

private fun ReadableMap.getOptionalMap(key: String): ReadableMap? {
    if (!hasKey(key) || isNull(key)) {
        return null
    }
    return getMap(key)
}

private fun ReadableMap.toStringMap(): Map<String, String> {
    val result = mutableMapOf<String, String>()
    val iterator = keySetIterator()
    while (iterator.hasNextKey()) {
        val key = iterator.nextKey()
        if (isNull(key)) {
            continue
        }
        val value = when (getType(key)) {
            ReadableType.String -> getString(key).orEmpty()
            ReadableType.Boolean -> getBoolean(key).toString()
            ReadableType.Number -> getDouble(key).toString()
            else -> null
        }
        if (value != null) {
            result[key] = value
        }
    }
    return result
}

private fun ReadableMap.toJSONObject(): JSONObject {
    val json = JSONObject()
    val iterator = keySetIterator()
    while (iterator.hasNextKey()) {
        val key = iterator.nextKey()
        json.put(key, toJSONValue(this, key))
    }
    return json
}

private fun ReadableArray.toJSONArray(): JSONArray {
    val json = JSONArray()
    for (index in 0 until size()) {
        json.put(toJSONValue(this, index))
    }
    return json
}

private fun toJSONValue(map: ReadableMap, key: String): Any? {
    if (map.isNull(key)) {
        return JSONObject.NULL
    }

    return when (map.getType(key)) {
        ReadableType.Null -> JSONObject.NULL
        ReadableType.Boolean -> map.getBoolean(key)
        ReadableType.Number -> numberValue(map.getDouble(key))
        ReadableType.String -> map.getString(key)
        ReadableType.Map -> map.getMap(key)?.toJSONObject()
        ReadableType.Array -> map.getArray(key)?.toJSONArray()
    }
}

private fun toJSONValue(array: ReadableArray, index: Int): Any? {
    if (array.isNull(index)) {
        return JSONObject.NULL
    }

    return when (array.getType(index)) {
        ReadableType.Null -> JSONObject.NULL
        ReadableType.Boolean -> array.getBoolean(index)
        ReadableType.Number -> numberValue(array.getDouble(index))
        ReadableType.String -> array.getString(index)
        ReadableType.Map -> array.getMap(index)?.toJSONObject()
        ReadableType.Array -> array.getArray(index)?.toJSONArray()
    }
}

private fun numberValue(value: Double): Number {
    val longValue = value.toLong()
    return if (value == longValue.toDouble()) {
        longValue
    } else {
        value
    }
}

File: ios/native/RNVNetworkDevModule.m
#import "RNVNetworkDevModule.h"

#import <TargetConditionals.h>
#import <UIKit/UIKit.h>

#import "RNVNetworkDevWebSocketClient.h"

static NSString * const RNVNetworkDevStatusEventName = @"RNVNetworkDevStatus";
static NSString * const RNVNetworkDevSDKVersion = @"0.1.0";

@interface RNVNetworkDevModule ()

@property (nonatomic, strong) RNVNetworkDevWebSocketClient *client;
@property (nonatomic, assign) BOOL hasListeners;
@property (nonatomic, assign) BOOL sdkEnabled;
@property (nonatomic, copy) NSString *sessionIdentifier;

@end

@implementation RNVNetworkDevModule

RCT_EXPORT_MODULE(RNVNetworkDevModule)

+ (BOOL)requiresMainQueueSetup {
    return NO;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _client = [RNVNetworkDevWebSocketClient new];
        _sdkEnabled = YES;
        _sessionIdentifier = [[NSUUID UUID] UUIDString];

        __weak typeof(self) weakSelf = self;
        _client.statusHandler = ^(NSString *state, NSString * _Nullable detail) {
            __strong typeof(weakSelf) strongSelf = weakSelf;
            if (strongSelf == nil) {
                return;
            }

            [strongSelf emitStatus:state detail:detail];
        };
    }

    return self;
}

- (NSArray<NSString *> *)supportedEvents {
    return @[RNVNetworkDevStatusEventName];
}

- (NSDictionary<NSString *, id> *)constantsToExport {
    return @{
        @"statusEventName": RNVNetworkDevStatusEventName,
        @"sdkVersion": RNVNetworkDevSDKVersion,
    };
}

RCT_EXPORT_METHOD(configure:(NSDictionary *)options) {
    self.sdkEnabled = [options[@"enabled"] respondsToSelector:@selector(boolValue)] ? [options[@"enabled"] boolValue] : YES;

    if (!self.sdkEnabled) {
        [self.client disconnectAndClearQueue];
        [self emitStatus:@"disabled" detail:@"SDK disabled by configuration."];
        return;
    }

    NSString *viewerURL = [options[@"viewerURL"] isKindOfClass:[NSString class]] ? options[@"viewerURL"] : nil;
    NSInteger maxQueueSize = [options[@"maxQueueSize"] respondsToSelector:@selector(integerValue)] ? [options[@"maxQueueSize"] integerValue] : 200;
    NSDictionary<NSString *, NSString *> *connectionHeaders = [options[@"connectionHeaders"] isKindOfClass:[NSDictionary class]] ? options[@"connectionHeaders"] : @{};

    [self.client configureWithURLString:viewerURL maxQueueSize:maxQueueSize connectionHeaders:connectionHeaders];
}

RCT_EXPORT_METHOD(setEnabled:(BOOL)enabled) {
    self.sdkEnabled = enabled;
    if (!enabled) {
        [self.client disconnectAndClearQueue];
        [self emitStatus:@"disabled" detail:@"SDK disabled at runtime."];
    }
}

RCT_EXPORT_METHOD(captureEvent:(NSDictionary *)event) {
    [self captureEvents:@[event]];
}

RCT_EXPORT_METHOD(captureEvents:(NSArray<NSDictionary *> *)events) {
    if (!self.sdkEnabled || events.count == 0) {
        return;
    }

    NSDictionary *envelope = [self buildEnvelopeWithEvents:events];
    [self.client enqueueEnvelope:envelope];
}

- (void)invalidate {
    self.client.statusHandler = nil;
    [self.client disconnectAndClearQueue];
}

#pragma mark - RCTEventEmitter

- (void)startObserving {
    self.hasListeners = YES;
}

- (void)stopObserving {
    self.hasListeners = NO;
}

#pragma mark - Helpers

- (NSDictionary *)buildEnvelopeWithEvents:(NSArray<NSDictionary *> *)events {
    return @{
        @"sdk": @{
            @"name": @"rnv_network_sdk_ios",
            @"version": RNVNetworkDevSDKVersion,
            @"schemaVersion": @1,
        },
        @"session": [self sessionMetadata],
        @"events": events,
    };
}

- (NSDictionary *)sessionMetadata {
    NSBundle *bundle = [NSBundle mainBundle];
    NSString *bundleIdentifier = bundle.bundleIdentifier ?: @"";
    NSString *displayName = [bundle objectForInfoDictionaryKey:@"CFBundleDisplayName"];
    if (displayName.length == 0) {
        displayName = [bundle objectForInfoDictionaryKey:@"CFBundleName"] ?: @"";
    }

    UIDevice *device = [UIDevice currentDevice];

    return @{
        @"id": self.sessionIdentifier,
        @"platform": @"ios",
        @"bundleIdentifier": bundleIdentifier,
        @"appName": displayName ?: @"",
        @"deviceName": device.name ?: @"",
        @"systemName": device.systemName ?: @"iOS",
        @"systemVersion": device.systemVersion ?: @"",
        @"isSimulator": @(RNVNetworkDevIsSimulator()),
    };
}

- (void)emitStatus:(NSString *)state detail:(nullable NSString *)detail {
    if (!self.hasListeners) {
        return;
    }

    dispatch_block_t emitBlock = ^{
        NSMutableDictionary *payload = [@{
            @"state": state ?: @"unknown",
            @"sessionId": self.sessionIdentifier ?: @"",
        } mutableCopy];

        if (detail.length > 0) {
            payload[@"detail"] = detail;
        }

        [self sendEventWithName:RNVNetworkDevStatusEventName body:payload];
    };

    if (NSThread.isMainThread) {
        emitBlock();
    } else {
        dispatch_async(dispatch_get_main_queue(), emitBlock);
    }
}

static BOOL RNVNetworkDevIsSimulator(void) {
#if TARGET_OS_SIMULATOR
    return YES;
#else
    return NO;
#endif
}

@end
